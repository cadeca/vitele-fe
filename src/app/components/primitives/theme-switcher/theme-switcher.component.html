<div class="theme-switcher">
  <div>
    <mat-icon
      class="icon"
      [matMenuTriggerFor]="menu">
      palette
    </mat-icon>
    <mat-menu #menu="matMenu">
      <button
        mat-menu-item
        (click)="changeThemeColor('orange-')">
        Orange
      </button>
      <button
        mat-menu-item
        (click)="changeThemeColor('magenta-')">
        Magenta
      </button>
    </mat-menu>
  </div>
  <div>
    <mat-slide-toggle (change)="changeThemeLight($event)" [checked]="darkThemeModeOn"></mat-slide-toggle>
  </div>
</div>
