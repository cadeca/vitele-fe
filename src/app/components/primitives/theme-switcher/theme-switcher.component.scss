@import "src/assets/style/variables";

/deep/ .mat-slide-toggle:not(.mat-disabled) {

  .mat-slide-toggle-bar {
    background-color: #fff;
    border: 1px solid #424242;
  }

  .mat-slide-toggle-thumb {
    background-color: #fff;
    height: 18px;
    width: 18px;
    border: 2px solid #424242;
  }

  &.mat-checked {

    .mat-slide-toggle-bar {
      background-color: #424242;
      border: 1px solid #fff;
    }

    .mat-slide-toggle-thumb {
      background-color: #424242;
      border: 2px solid #fff;
    }
  }
}

.theme-switcher {
  display: flex;
  flex-direction: column;

  > div {
    width: 50%;
    display: flex;

    &:hover {
      cursor: pointer;
    }

    .icon {
      padding-left: 6px;
      width: 100%;
    }
  }
}

/deep/ .mat-slide-toggle {
  height: 50px !important;
  transform: rotate(90deg);

  label {
    margin: 0;
  }
}
