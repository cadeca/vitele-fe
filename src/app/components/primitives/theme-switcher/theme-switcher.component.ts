import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'wl-theme-switcher',
  templateUrl: './theme-switcher.component.html',
  styleUrls: ['./theme-switcher.component.scss']
})
export class ThemeSwitcherComponent implements OnInit {

  themeLight = 'light-theme';
  themeColor = 'magenta';
  darkThemeModeOn = false;
  constructor() { }

  ngOnInit(): void {
    this.setDefaultTheme();
  }

  setDefaultTheme(): void {
    // if theme is stored in storage - use it
    if (localStorage.getItem('pxThemeLight') && localStorage.getItem('pxThemeColor')){
      // set theme color to one from storage
      this.themeColor = localStorage.getItem('pxThemeColor');
      this.themeLight = localStorage.getItem('pxThemeLight');
      // add that class to body
      const body = document.getElementsByTagName('body')[0];
      body.classList.add(this.themeColor + this.themeLight);
    }
  }

  changeTheme(themeColorToSet, themeLightToSet): void {
    const body = document.getElementsByTagName('body')[0];
    body.classList.remove(this.themeColor + this.themeLight);

    // switch theme
    if (this.themeColor !== themeColorToSet && themeColorToSet) {
      this.themeColor = themeColorToSet;
    }
    if (this.themeLight !== themeLightToSet && themeLightToSet) {
      this.themeLight = themeLightToSet;
    }

    body.classList.add(this.themeColor + this.themeLight);

    // save it to local storage

    localStorage.setItem('pxThemeColor', this.themeColor);
    localStorage.setItem('pxThemeLight', this.themeLight);
  }

  changeThemeLight(event): void {
    if (event.checked) {
      this.darkThemeModeOn = true;
      this.changeTheme(null, 'dark-theme');
    } else {
      this.darkThemeModeOn = false;
      this.changeTheme(null, 'light-theme');
    }
  }

  changeThemeColor(color: string): void {
    this.changeTheme(color, null);
  }
}
