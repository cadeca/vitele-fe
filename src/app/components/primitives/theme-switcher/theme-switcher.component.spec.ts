import {ThemeSwitcherComponent} from './theme-switcher.component';
import {async, fakeAsync, TestBed, tick} from '@angular/core/testing';
import {MaterialModule} from '../../../material.module';

const DARK_THEME = 'dark-theme';
const LIGHT_THEME = 'light-theme';
const ORANGE_COLOR = 'orange-';

describe('ThemeSwitcherComponent', () => {
  let fixture;
  let component;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        MaterialModule
      ],
      declarations: [ ThemeSwitcherComponent ]
    })
      .compileComponents().then(() => {
        fixture = TestBed.createComponent(ThemeSwitcherComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
    });
  }));

  it('should check initial state', () => {
    expect(component.themeLight).toBe(LIGHT_THEME);
    expect(component.themeColor).toBe(ORANGE_COLOR);
    expect(component.darkThemeModeOn).toBe(false);
  });

  it('#dark-theme-toggle should trigger the correct method', fakeAsync (() => {
    expect(component.themeLight).toBe(LIGHT_THEME);
    spyOn(component, 'changeThemeLight');

    const input = fixture.debugElement.nativeElement.querySelector('#dark-theme-toggle');
    input.dispatchEvent(new Event('change', {target: {checked: true}} as any));
    tick();

    expect(component.changeThemeLight).toHaveBeenCalled();
  }));

  it('toggle dark theme should alter the correct attributes', () => {
    expect(component.themeLight).toBe(LIGHT_THEME);
    expect(localStorage.getItem('pxThemeLight')).toBe(LIGHT_THEME);

    component.changeThemeLight({checked: true});
    fixture.detectChanges();

    expect(component.themeLight).toBe(DARK_THEME);
    expect(localStorage.getItem('pxThemeLight')).toBe(DARK_THEME);
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
