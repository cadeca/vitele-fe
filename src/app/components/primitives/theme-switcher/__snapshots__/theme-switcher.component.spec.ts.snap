// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcherComponent should match snapshot 1`] = `
<wl-theme-switcher
  darkThemeModeOn={[Function Boolean]}
  themeColor={[Function String]}
  themeLight={[Function String]}
>
  <div
    class="theme-switcher"
  >
    <div>
      <mat-icon
        aria-haspopup="true"
        aria-hidden="true"
        class="icon mat-icon notranslate mat-menu-trigger material-icons mat-icon-no-color"
        id="color-change-trigger"
        role="img"
      >
         palette 
      </mat-icon>
      <mat-menu>
        
      </mat-menu>
    </div>
    <div>
      <mat-slide-toggle
        class="mat-slide-toggle mat-accent mat-checked"
        id="dark-theme-toggle"
        tabindex="-1"
      >
        <label
          class="mat-slide-toggle-label"
          for="dark-theme-toggle-input"
        >
          <div
            class="mat-slide-toggle-bar mat-slide-toggle-bar-no-side-margin"
          >
            <input
              aria-checked="true"
              class="mat-slide-toggle-input cdk-visually-hidden"
              id="dark-theme-toggle-input"
              role="switch"
              tabindex="0"
              type="checkbox"
            />
            <div
              class="mat-slide-toggle-thumb-container"
            >
              <div
                class="mat-slide-toggle-thumb"
              />
              <div
                class="mat-slide-toggle-ripple mat-focus-indicator mat-ripple"
                mat-ripple=""
              >
                <div
                  class="mat-ripple-element mat-slide-toggle-persistent-ripple"
                />
              </div>
            </div>
          </div>
          <span
            class="mat-slide-toggle-content"
          >
            <span
              style="display:none"
            >
              Â 
            </span>
          </span>
        </label>
      </mat-slide-toggle>
    </div>
  </div>
</wl-theme-switcher>
`;
